opengrok-indexer/src/test/java/org/opengrok/indexer/util/ProgressTest.java

```diff
@@ -0,0 +1,90 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+import static org.junit.Assert.assertNotNull;
+import static org.mockito.ArgumentMatchers.*;
+import static org.mockito.Mockito.atLeast;
+import static org.mockito.Mockito.times;
+
+public class ProgressTest {
+    @BeforeClass
+    public static void setup() {
+        // needed to spawn logger thread in Progress
+        RuntimeEnvironment.getInstance().setPrintProgress(true);
+    }
+
+    @Test
+    public void testProgress() throws InterruptedException {
+        final Logger logger = Mockito.mock(Logger.class);
+        final int totalCount = 100; // use int because of Mockito.calls()/times() only accepts int
+
+        Mockito.when(logger.isLoggable(any())).thenReturn(true);
+        try (final Progress progress = new Progress(logger, "foo", totalCount)) {
+            assertNotNull(progress.getLoggerThread());
+
+            // Progress does not give any indication that the logger thread reached the point when
+            // it can accept notifications from increment() so check the Thread state.
+            while (progress.getLoggerThread().getState() != Thread.State.WAITING) {
+                System.out.println("Waiting for the logger thread to reach the initial wait()");
+                TimeUnit.MILLISECONDS.sleep(10);
+            }
+
+            for (int i = 0; i < totalCount; i++) {
+                progress.increment();
+                // Give the logger thread some time to log.
+                while (progress.getLoggerThread().getState() != Thread.State.WAITING) {
+                    TimeUnit.MILLISECONDS.sleep(10);
+                }
+            }
+        }
+
+        Mockito.verify(logger, times(totalCount)).log(any(), anyString(), any(Object[].class));
+    }
+
+    @Test
+    public void testThreads() throws InterruptedException {
+        final Logger logger = Mockito.mock(Logger.class);
+        final int totalCount = Runtime.getRuntime().availableProcessors();
+        List<Future<?>> futures = new ArrayList<>();
+
+        Mockito.when(logger.isLoggable(any())).thenReturn(true);
+        ExecutorService executor = Executors.newFixedThreadPool(totalCount);
+        System.out.println(String.format("Will run %d threads", totalCount));
+        try (final Progress progress = new Progress(logger, "foo", totalCount)) {
+            while (progress.getLoggerThread().getState() != Thread.State.WAITING) {
+                System.out.println("Waiting for the logger thread to reach the initial wait()");
+                TimeUnit.MILLISECONDS.sleep(10);
+            }
+
+            for (int i = 0; i < totalCount; i++) {
+                futures.add(executor.submit(progress::increment));
+            }
+        }
+
+        System.out.println("Waiting for threads to finish");
+        futures.forEach(future -> {
+            try {
+                future.get();
```

Comment:
Might the `AutoCloseable` `close()` have already finished? Should the futures waits happen inside the `try`?

Reply:
yes, that would be more robust.